### DISCUSSION TITLE
How to persist WorkflowInput values after executing Bookmark Callbacks

### DISCUSSION THREAD

**Discussion by defunky** [elsa-workflows/elsa-core#5292] [Q&A]
I have a workflow with some values in WorkflowInput, one of these activities I have also assigns a new value within to WorkflowInput dictonary via the `ExecuteAsync` using `ActivityExecutionContext`.  After the assignment I'm creating a boomark with a callback method, however upon calling the `ResumeWorkflowsAsync` on IWorkflowRunTime, the WorkflowInput is empty and any subsequent activites that rely on values in WorkflowInput fail now as the previously assign values are gone.

I can see that `TriggerWorkflowsOptions` has the option to pass in WorkflowInput when doing a Resume but I don't want to reassign all the values that were assigned before. Is there a way to persist the data after the callback and is it possible to append WorkflowInput via TriggerWorkflowsOptions instead of completely overwriting it?

**Comment by glime-ai** [comment]
To persist data across workflow execution, including after resuming from a bookmark, you should ensure that the data is being saved to the workflow instance's state. When you use `ActivityExecutionContext` to set values, these should be stored in the workflow instance and be available upon resumption, provided that the workflow instance itself is being persisted to a durable store (like a database).

If you are experiencing issues where the `WorkflowInput` is empty after resuming, it could be due to a few reasons:

1. The workflow instance state is not being persisted correctly.
2. The data is not being set in a way that it is included in the workflow instance state.
3. There might be a bug or a missing feature in the version of Elsa you are using.

Regarding appending to `WorkflowInput` via `TriggerWorkflowsOptions` without overwriting it, the current behavior of Elsa seems to be that it replaces the entire input dictionary. However, you could potentially work around this by manually merging the existing inputs with the new inputs before resuming the workflow.

Here's a general approach to ensure data persistence and appending inputs:

1. Make sure you are using a persistent store for workflow instances.
2. When setting data that needs to persist, ensure it is being added to the workflow instance's state.
3. Before resuming a workflow, retrieve the current workflow instance and its inputs.
4. Merge the existing inputs with the new inputs you want to add.
5. Pass the merged inputs to `TriggerWorkflowsOptions`.

If you are still facing issues, please provide more details about the version of Elsa you are using and the exact steps you are taking to set and retrieve the workflow input. This will help in giving a more precise solution.

For more information on how to work with inputs and outputs in Elsa, you can refer to the documentation on input and output: https://v3.elsaworkflows.io/docs/core-concepts/input-output

If you believe you are encountering a bug or a limitation in Elsa, you may want to check the GitHub issues or discussions for similar problems or report a new issue: https://github.com/elsa-workflows/elsa-core/issues

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).

**Comment by defunky** [comment]
The workflow is being persisted on a database, if I do the same activity without any bookmarks the `WorkflowInput `persists to the next activity. Once I add the Bookmark with a callback `ActivityExecutionContext `in the callback function the `WorkflowInput `is empty even though my `TriggerWorkflowsOptions `is not being set on `ResumeWorkflowsAsync`.

**Comment by boltth** [comment]
@defunky Did you ever get an answer or find a solution to this? I'm experiencing the same.

**Reply by defunky** [reply]
It's intended to be like this apparently, if you want persistence you have to bind it to a workflow variable on the context

**Reply by boltth** [reply]
Thank you @defunky . I just recognized that I need to define the inputs on the workflow (InputDefinition) including setting a StorageDriverType. Now it works like a charm. I think that's exactly what you mean.

**Comment by sfmskywalker** [comment]
Closing as resolved. The persisted workflow state only includes workflow inputs whose input definitions use a persistent storage driver. In Elsa 3.0.0, that is `WorkflowStorageDriver`; values added only to the runtime input dictionary are not persisted across bookmark resumption.

For values that must survive a bookmark, define the workflow input with the appropriate storage driver, or store the value in a workflow variable configured for persistent storage. This matches the resolution reported above.

Relevant source: [`WorkflowStateExtractor.GetPersistableInput`](https://github.com/elsa-workflows/elsa-core/blob/3.0.0/src/modules/Elsa.Workflows.Core/Services/WorkflowStateExtractor.cs#L39-L52) and [`InputDefinition.StorageDriverType`](https://github.com/elsa-workflows/elsa-core/blob/3.0.0/src/modules/Elsa.Workflows.Core/Models/InputDefinition.cs#L6-L17).
